crates/sui-adapter/src/programmable_transactions/linkage_view.rs

```diff
@@ -0,0 +1,125 @@
+    resolver::{LinkageResolver, ModuleResolver, ResourceResolver},
+};
+use sui_types::{
+    base_types::{ObjectID, ObjectRef},
+    error::{SuiError, SuiResult},
+    event::Event,
+    move_package::UpgradeInfo,
+    object::Object,
+    storage::{BackingPackageStore, ChildObjectResolver, ObjectChange, ParentSync, Storage},
+};
+
+use super::types::StorageView;
+
+/// TODO Docs
+#[allow(dead_code)]
+pub struct LinkageView<'state, S: StorageView> {
+    state_view: &'state S,
+    linkage_info: Option<LinkageInfo>,
+}
+
+/// TODO Docs
+#[allow(dead_code)]
+pub struct LinkageInfo {
+    link_context: ObjectID,
+    linkage_table: BTreeMap<ObjectID, UpgradeInfo>,
+}
+
+impl<'state, S: StorageView> LinkageView<'state, S> {
+    pub fn new(state_view: &'state S) -> Self {
+        Self {
+            state_view,
+            linkage_info: None,
+        }
+    }
+
+    pub fn storage(&self) -> &'state S {
+        self.state_view
+    }
+}
+
+impl<'state, S: StorageView> LinkageResolver for LinkageView<'state, S> {
+    type Error = SuiError;
+
+    fn link_context(&self) -> AccountAddress {
+        AccountAddress::ZERO
+    }
+
+    fn relocate(&self, module_id: &ModuleId) -> Result<ModuleId, Self::Error> {
+        Ok(module_id.clone())
+    }
+
+    fn defining_module(
+        &self,
+        module_id: &ModuleId,
+        _struct: &move_core_types::identifier::IdentStr,
+    ) -> Result<ModuleId, Self::Error> {
+        Ok(module_id.clone())
+    }
+}
+
+/** Remaining implementations delegated to StorageView ************************/
+
+impl<'state, S: StorageView> ResourceResolver for LinkageView<'state, S> {
+    type Error = <S as ResourceResolver>::Error;
+
+    fn get_resource(
+        &self,
+        address: &AccountAddress,
+        typ: &StructTag,
+    ) -> Result<Option<Vec<u8>>, Self::Error> {
+        self.state_view.get_resource(address, typ)
+    }
+}
+
+impl<'state, S: StorageView> ModuleResolver for LinkageView<'state, S> {
+    type Error = <S as ModuleResolver>::Error;
+
+    fn get_module(&self, id: &ModuleId) -> Result<Option<Vec<u8>>, Self::Error> {
+        self.state_view.get_module(id)
+    }
+}
+
+impl<'state, S: StorageView> BackingPackageStore for LinkageView<'state, S> {
+    fn get_package_object(&self, package_id: &ObjectID) -> SuiResult<Option<Object>> {
+        self.state_view.get_package_object(package_id)
+    }
+}
+
+impl<'state, S: StorageView> Storage for LinkageView<'state, S> {
+    fn read_object(&self, id: &ObjectID) -> Option<&Object> {
+        self.state_view.read_object(id)
+    }
+
+    fn reset(&mut self) {
+        unimplemented!("Read-only storage only.")
+    }
+
+    fn log_event(&mut self, _event: Event) {
+        unimplemented!("Read-only storage only.")
+    }
+
+    fn apply_object_changes(&mut self, _changes: BTreeMap<ObjectID, ObjectChange>) {
+        unimplemented!("Read-only storage only.")
+    }
+}
+
+impl<'state, S: StorageView> ParentSync for LinkageView<'state, S> {
```

Comment:
Is `LinkageView` actually used as `StorageView`? If not, then technically we probably don't have to implement `ParentSync` (nor perhaps `BackingPackageStore`) for it.

Reply:
Never mind. I can see that `LinkageView` is passed to `Session` as `StorageView`

Reply:
No, I think you might be right there -- I'll take a look!